parliament/community_auditors/privilege_escalation.py

```diff
@@ -0,0 +1,109 @@
+from policy_sentry.shared.database import connect_db
+from policy_sentry.util.policy_files import get_actions_from_policy
+from policy_sentry.analysis.analyze import determine_actions_to_expand
+
+
+def audit(policy):
+    db_session = connect_db('bundled')
+    actions_in_policy = get_actions_from_policy(policy.policy_json)
+    expanded_actions = determine_actions_to_expand(db_session, actions_in_policy)
+    permissions_on_other_users(policy, expanded_actions)
+
+
+# Categories based on https://know.bishopfox.com/blog/5-privesc-attack-vectors-in-aws
+
+def permissions_on_other_users(policy, expanded_actions):
+    escalation_methods = {
+        # 1. IAM Permissions on Other Users
+        'CreateAccessKey': [
+            'iam:createaccesskey'
+        ],
+        'CreateLoginProfile': [
+            'iam:createloginprofile'
+        ],
+        'UpdateLoginProfile': [
+            'iam:updateloginprofile'
+        ],
+        # 2. Permissions on Policies
+        'CreateNewPolicyVersion': [
+            'iam:createpolicyversion'
+        ],
+        'SetExistingDefaultPolicyVersion': [
+            'iam:setdefaultpolicyversion'
+        ],
+        'AttachUserPolicy': [
+            'iam:attachuserpolicy'
+        ],
+        'AttachGroupPolicy': [
+            'iam:attachgrouppolicy'
+        ],
+        'AttachRolePolicy': [
+            'iam:attachrolepolicy',
+            'sts:assumerole'
+        ],
+        'PutUserPolicy': [
+            'iam:putuserpolicy'
+        ],
+        'PutGroupPolicy': [
+            'iam:putgrouppolicy'
+        ],
+        'PutRolePolicy': [
+            'iam:putrolepolicy',
+            'sts:assumerole'
+        ],
+        'AddUserToGroup': [
+            'iam:addusertogroup'
+        ],
+        # 3. Updating an AssumeRolePolicy
+        'UpdateRolePolicyToAssumeIt': [
+            'iam:updateassumerolepolicy',
+            'sts:assumerole'
+        ],
+        # 4. iam:PassRole:*
+        'CreateEC2WithExistingIP': [
+            'iam:passrole',
+            'ec2:runinstances'
+        ],
+        'PassExistingRoleToNewLambdaThenInvoke': [
+            'iam:passrole',
+            'lambda:createfunction',
+            'lambda:invokefunction'
+        ],
+        'PassExistingRoleToNewLambdaThenTriggerWithNewDynamo': [
```

Comment:
I'm not a fan of how these detections look at specific combinations of privileges because they miss a lot. For example, you can trigger a Lambda from an S3 bucket, which is roughly the same as `PassExistingRoleToNewLambdaThenTriggerWithNewDynamo `, but would not be detected with this rule set. This is too much of a cat and mouse game.

I strongly believe that one should simply look for any `iam` privileges that aren't `List` or `Describe`.  That avoids the whole cat and mouse game. 

One could argue that if all you had was `iam:passrole` without any other privileges, then that wouldn't have a privilege escalation path, and so that would be an FP, but that would also be odd and worth flagging as a finding. One could also argue that this list better defines what the privilege escalations are, and there are examples documented by Rhino, but I'm ok letting the user figure out why this might be bad.  Because even in these examples, the user has to have a pretty decent understanding of IAM to understand these privilege escalation paths.

Reply:
* *"I'm not a fan of how these detections look at specific combinations of privileges because they miss a lot. For example, you can trigger a Lambda from an S3 bucket, which is roughly the same as PassExistingRoleToNewLambdaThenTriggerWithNewDynamo , but would not be detected with this rule set. This is too much of a cat and mouse game."*

I agree that it wouldn't detect that - but I'd counter that the first case - `PassExistingRoleToNewLambdaThenTriggerWithNewDynamo` - **confirms** that privilege escalation path, whereas triggering Lambda from an S3 bucket (which would only require `iam:passrole`, `lambda:createfunction`, then give the lambda role `s3:getobject`) would require additional triaging.  If we don't allow the combo-based privesc, it forces all findings into the second case, rather than at least allowing for some combos to be confirmed in the first. Does that make sense?

Reply:
* *"I strongly believe that one should simply look for any iam privileges that aren't List or Describe. That avoids the whole cat and mouse game."* 

See my first comment that proposes "IAM Tampering" + "Privilege Escalation."